### ISSUE TITLE
cunit tests depending on abstraction leaks

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#3178]
There's a new make target `make check-discrete` on the master branch, which runs every test in isolation.

All our tests work when run normally via `make check`.

`make check-discrete` finds 55 tests that fail when run in isolation, which are probably depending on some state leak between tests that doesn't exist when they're run alone.

**Comment by elliefm** [comment]
WIP here https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v33/3178-cunit-config-leaks

All 55 failing tests depend on the config state being leaked between tests.

I've fixed up the config leaks, so that these tests now fail for a normal run too.  I've then fixed up most of the failing test suites to initialise config themselves, so they're no longer dependent on it leaking.

The remaining exception (which now fails for a normal "make check"!) is mboxname.testc, which both depends on config leaking in from elsewhere, AND wants to mess around with it a lot internally.  I haven't yet figured out a good approach for fixing it yet.
